File: labs/14-prompt-first-dev/node/main.ts
import fs from "fs";
import path from "path";
import { OpenAIClient } from "../../../shared/config/openai_client";

const REQUIREMENT =
  "Build an API for a task manager. Users can create a task with a title and " +
  "optional description, list all tasks, and mark a task as complete. " +
  "Store tasks in memory only.";

const PROMPT_VERSION = process.env.PROMPT_VERSION || "v2";

function loadPromptTemplate(version: string): string {
  const promptPath = path.join(__dirname, "prompts", version, "spec_gen.txt");
  if (!fs.existsSync(promptPath)) {
    const promptDir = path.join(__dirname, "prompts");
    const available = fs.existsSync(promptDir)
      ? fs
          .readdirSync(promptDir)
          .filter((p) => fs.statSync(path.join(promptDir, p)).isDirectory())
      : [];
    throw new Error(
      `Prompt template not found for version '${version}'. Available versions: ${available.join(", ") || "none"}`
    );
  }
  return fs.readFileSync(promptPath, "utf8");
}

function renderPrompt(template: string, requirement: string): string {
  return template.replace(/\{\{\s*requirement\s*\}\}/g, requirement);
}

function extractCodeBlocks(content: string): [string, string] {
  const mdMatch = content.match(/```markdown\s*\n([\s\S]*?)\n```/);
  const pyMatch = content.match(/```python\s*\n([\s\S]*?)\n```/);
  if (!mdMatch || !pyMatch) {
    throw new Error(
      "Model response did not contain both required markdown and python code blocks."
    );
  }
  return [mdMatch[1].trim(), pyMatch[1].trim()];
}

async function generateArtifacts(
  client: OpenAIClient,
  requirement: string,
  version: string
): Promise<[string, string]> {
  const template = loadPromptTemplate(version);
  const prompt = renderPrompt(template, requirement);

  console.log(`Using prompt version: ${version}`);
  console.log(`Sending ${prompt.length} prompt characters to model.`);

  const response = await client.chatCompletion({
    messages: [{ role: "user", content: prompt }],
    temperature: 0.2,
    max_tokens: 1500,
  });

  const content = response.choices[0].message.content;
  const [spec, scaffold] = extractCodeBlocks(content);
  return [spec, scaffold];
}

function writeArtifacts(spec: string, scaffold: string, outputDir: string): void {
  fs.mkdirSync(outputDir, { recursive: true });
  const specPath = path.join(outputDir, "api_spec.md");
  const scaffoldPath = path.join(outputDir, "scaffold.py");
  fs.writeFileSync(specPath, spec + "\n", "utf8");
  fs.writeFileSync(scaffoldPath, scaffold + "\n", "utf8");
  console.log(`Wrote artifacts to ${outputDir}`);
  console.log(`  - ${path.basename(specPath)}`);
  console.log(`  - ${path.basename(scaffoldPath)}`);
}

function preview(text: string, maxChars = 400): string {
  if (text.length <= maxChars) return text;
  return text.slice(0, maxChars).trimEnd() + "\n...";
}

async function main() {
  const client = new OpenAIClient();

  let spec: string;
  let scaffold: string;
  try {
    [spec, scaffold] = await generateArtifacts(client, REQUIREMENT, PROMPT_VERSION);
  } catch (exc: any) {
    console.error("Failed to generate artifacts:", exc.message);
    process.exit(1);
  }

  const outputDir = path.join(__dirname, "generated");
  try {
    writeArtifacts(spec, scaffold, outputDir);
  } catch (exc: any) {
    console.error("Failed to write artifacts:", exc.message);
    process.exit(1);
  }

  console.log("\n=== Generated API Spec Preview ===");
  console.log(preview(spec));
  console.log("\n=== Generated Scaffold Preview ===");
  console.log(preview(scaffold));
  console.log(`\nArtifacts written to: ${path.resolve(outputDir)}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: labs/14-prompt-first-dev/python/main.py
"""Lab 14: Prompt-first development.

Loads a versioned prompt template, renders it with a requirement, asks an LLM
to produce an API spec and a FastAPI scaffold, then writes the artifacts to disk.
"""

import json
import logging
import os
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging

logger = logging.getLogger("agent-labs")

REQUIREMENT = (
    "Build an API for a task manager. Users can create a task with a title and "
    "optional description, list all tasks, and mark a task as complete. "
    "Store tasks in memory only."
)

PROMPT_VERSION = os.getenv("PROMPT_VERSION", "v2")


def load_prompt_template(version: str) -> str:
    prompt_path = Path(__file__).parent / "prompts" / version / "spec_gen.txt"
    if not prompt_path.exists():
        available = [p.name for p in (Path(__file__).parent / "prompts").iterdir() if p.is_dir()]
        raise FileNotFoundError(
            f"Prompt template not found for version '{version}'. "
            f"Available versions: {available or 'none'}"
        )
    return prompt_path.read_text(encoding="utf-8")


def render_prompt(template: str, requirement: str) -> str:
    return template.replace("{{ requirement }}", requirement)


def extract_code_blocks(content: str) -> tuple[str, str]:
    """Extract the first markdown and first python code blocks."""
    md_match = re.search(r"```markdown\s*\n(.*?)\n```", content, re.DOTALL)
    py_match = re.search(r"```python\s*\n(.*?)\n```", content, re.DOTALL)

    if not md_match or not py_match:
        raise ValueError("Model response did not contain both required markdown and python code blocks.")

    return md_match.group(1).strip(), py_match.group(1).strip()


def generate_artifacts(client: OpenAIClient, requirement: str, version: str) -> tuple[str, str]:
    template = load_prompt_template(version)
    prompt = render_prompt(template, requirement)

    logger.info("Using prompt version: %s", version)
    logger.info("Sending %d prompt characters to model.", len(prompt))

    response = client.chat_completion(
        messages=[{"role": "user", "content": prompt}],
        temperature=0.2,
        max_tokens=1500,
    )

    content = response["choices"][0]["message"]["content"]
    spec, scaffold = extract_code_blocks(content)
    return spec, scaffold


def write_artifacts(spec: str, scaffold: str, output_dir: Path) -> None:
    output_dir.mkdir(parents=True, exist_ok=True)

    spec_path = output_dir / "api_spec.md"
    scaffold_path = output_dir / "scaffold.py"

    spec_path.write_text(spec + "\n", encoding="utf-8")
    scaffold_path.write_text(scaffold + "\n", encoding="utf-8")

    logger.info("Wrote artifacts to %s", output_dir)
    logger.info("  - %s", spec_path.name)
    logger.info("  - %s", scaffold_path.name)


def preview(text: str, max_chars: int = 400) -> str:
    if len(text) <= max_chars:
        return text
    return text[:max_chars].rstrip() + "\n..."


def main():
    configure_logging()
    client = OpenAIClient()

    try:
        spec, scaffold = generate_artifacts(client, REQUIREMENT, PROMPT_VERSION)
    except Exception as exc:
        logger.error("Failed to generate artifacts: %s", exc)
        sys.exit(1)

    output_dir = Path(__file__).parent / "generated"
    try:
        write_artifacts(spec, scaffold, output_dir)
    except OSError as exc:
        logger.error("Failed to write artifacts: %s", exc)
        sys.exit(1)

    print("\n=== Generated API Spec Preview ===")
    print(preview(spec))
    print("\n=== Generated Scaffold Preview ===")
    print(preview(scaffold))
    print(f"\nArtifacts written to: {output_dir.resolve()}")


if __name__ == "__main__":
    main()
